### PR TITLE
Copy over runfiles library from Bazel

### PR DESCRIPTION
**Pull request by fmeum**
The runfiles library can be maintained independently of Bazel releases and `bazel_tools` can refer to it via an alias.

Also set flags to build and test with a hermetic JDK 8 to ensure compatibility with that version.

### CODE REVIEW

**Comment by fmeum** [comment]
@hvadehra @meteorcloudy I just realized that we could also move this over and delete it from bazel_tools. What do you think? We could fix https://github.com/bazelbuild/bazel/issues/24150 here.

**Comment by comius** [comment]
The move sounds fine to me. I'll let @hvadehra do the review.

File: java/runfiles/src/main/java/com/google/devtools/build/runfiles/AutoBazelRepositoryProcessor.java
Comment (fzakaria) [inline java/runfiles/src/main/java/com/google/devtools/build/runfiles/AutoBazelRepositoryProcessor.java:98]: This is the https://github.com/bazelbuild/bazel/issues/24150

**Comment by fzakaria** [comment]
> @hvadehra @meteorcloudy I just realized that we could also move this over and delete it from bazel_tools. What do you think? We could fix [bazelbuild/bazel#24150](https://github.com/bazelbuild/bazel/issues/24150) here.

I included the fix in https://github.com/bazelbuild/bazel/pull/24161
It doesn't have any unit tests since there seem to be none to add.
A further contribution may be to include https://github.com/google/compile-testing at some point

**Comment by fmeum** [comment]
Stacked on https://github.com/bazelbuild/rules_java/pull/238

**Comment by fmeum** [comment]
@hvadehra This is ready for review now, CI is green.

**Comment by fmeum** [comment]
@hvadehra Friendly ping

File: .bazelrc
Comment (hvadehra) [inline .bazelrc:3]: Are these needed for the rest of this change to work? If not, WDYT about a separate CI job with these flags?

File: test/repositories.bzl
Comment (hvadehra) [inline test/repositories.bzl:4]: Any reason to not use http_jar from @rules_java itself? I think you said as much in https://github.com/bazelbuild/rules_java/pull/238

**Review by hvadehra — COMMENTED** [review]
Sorry about the delay.

File: test/repositories.bzl
Comment (fmeum) [reply test/repositories.bzl:4]: Will update, that change just hadn't been merged yet and I didn't want to block this PR on it.

File: .bazelrc
Comment (fmeum) [reply .bazelrc:3]: Without this, running `bazel test //...` locally would fail if the local JDK is more recent than the JDK in the Java compilation toolchain, which is very likely (I have JDK 23 locally). 

Do you see a downside to setting these flags? Ensuring that CI runs and local runs see the same result seems important for ease of contributing. This wasn't a problem prior to this change as the repo didn't actually contain any Java source files.

Comment (hvadehra) [reply .bazelrc:3]: Yeah, that makes sense, but if it's not too much trouble, doing it in a separate change would be nicer.

Comment (fmeum) [reply .bazelrc]: I split out https://github.com/bazelbuild/rules_java/pull/241

**Comment by fmeum** [comment]
@Wyverald `http_jar` in `rules_java` doesn't work with Bazel 6 as it loads `cache.bzl`, which isn't available in that Bazel version.

File: test/repositories.bzl
Comment (fmeum) [reply test/repositories.bzl:6]: Added a comment to explain why.

**Comment by Wyverald** [comment]
> @Wyverald `http_jar` in `rules_java` doesn't work with Bazel 6 as it loads `cache.bzl`, which isn't available in that Bazel version.

I'm aware of this (decided not to try to fix it since it's too much work). Is it blocking you?

**Comment by fmeum** [comment]
If you are aware of it's fine, it just means that rules_java can't use it yet.

File: test/repositories.bzl
Comment (hvadehra) [reply test/repositories.bzl:7]: It should be possible to use via the compatibility proxy layer (redirecting to `@bazel_tools` instead for older Bazel). But there's no need to block this PR on that and can happen later.

**Review by hvadehra — DISMISSED** [review]

**Comment by hvadehra** [comment]
@fmeum This is causing a few issues on import, could you please change the newly added `BUILD` files to `BUILD.bazel`?

**Comment by fmeum** [comment]
@hvadehra Done

**Comment by fmeum** [comment]
@hvadehra Could you cut a new release so that this can be integrated into https://github.com/bazelbuild/bazel/pull/24219?

**Comment by hvadehra** [comment]
Done: https://github.com/bazelbuild/bazel-central-registry/pull/3213

**Pull request closed without merging**
